File: lab4/java_solution/highlightingtext/Document.java
// Document.java
package highlightingtext;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.io.PrintStream;
import java.util.ArrayList;
import java.util.List;

public class Document {
    private static final int DEFAULT_BUFFER_SIZE = 2048;
    private ArrayList<Format.CharColor> formatChars = new ArrayList<>(DEFAULT_BUFFER_SIZE);
    private List<Character> chstr = new ArrayList<>(DEFAULT_BUFFER_SIZE);

    public Document(String path) {
        try (BufferedReader reader = new BufferedReader(new FileReader(path))) {
            int ch;
            while ((ch = reader.read()) != -1) {
                chstr.add((char) ch);
                formatChars.add(Format.CharColor.WHITE);
            }
        } catch (IOException e) {
            throw new IllegalArgumentException("file did not open.", e);
        }
    }


    public Document append(String str)
    {
        for (Character c : str.toCharArray()) {
            this.chstr.add(c);   
            this.formatChars.add(Format.CharColor.WHITE);
        }
        return this;
    }

    public Document append(String str, Format.CharColor cl)
    {
        for (Character c : str.toCharArray()) {
            this.chstr.add(c);   
            this.formatChars.add(cl);
        }
        return this;
    }

    public void outDocument(PrintStream os) {
        Format.SetConsoleColor setConsoleColor = new Format.SetConsoleColor(os);
        for (int i = 0; i < chstr.size(); i++) {
            setConsoleColor.setColor(formatChars.get(i));
            os.print(chstr.get(i));
            setConsoleColor.setColor(Format.CharColor.WHITE);
        }
    }

    public void setFormat(int begin, int end, Format.CharColor cl) {
        for (int i = begin; i < end; i++) {
            formatChars.set(i,cl);
        }
    }
    
    public List<Character> getChstr() {
        return chstr;
    }
}

File: lab6_7/cpp_solution/Document.cpp
#include "Document.hpp"

std::ostream& operator<<(std::ostream& os, const Document& d)
{
    Format::SetConsoleColor setConsoleColor(os);
    for (size_t i = 0; i < d.chstr.size(); ++i)
    {
        setConsoleColor(static_cast<int>(d.formatChars[i]), 0);
        os << d.chstr[i];
        setConsoleColor(static_cast<int>(Format::charColor::White), 0);
    }
    return os;
}

Document operator+(const Document& lhs, const Document& rhs)
{
    Document res = lhs;
    try 
    {
        res.formatChars.reserve(lhs.formatChars.capacity() + rhs.formatChars.capacity());
        res.chstr.reserve(lhs.chstr.capacity() + rhs.chstr.capacity());
    } 
    catch(std::length_error& len_e)
    {
        res.formatChars.reserve(lhs.formatChars.size() + rhs.formatChars.size());
        res.chstr.reserve(lhs.chstr.size() + rhs.chstr.size());
    }

    res.formatChars.insert(res.formatChars.end(), rhs.formatChars.begin(), rhs.formatChars.end());
    res.chstr.insert(res.chstr.end(), rhs.chstr.begin(), rhs.chstr.end());
    return res;
}


void Document::setFormat(int begin, int end, Format::charColor cl)
{
    for (int i = begin; i < end; ++i)
        formatChars[i] = cl;
}

void Document::insertChar(std::vector<char>::const_iterator pos, char ch)
{
    chstr.insert(pos, ch);
}

Document::Document(const std::string& path)
{
    chstr.reserve(DocumentConst::DEFAULT_BUFFER_SIZE);
    formatChars.reserve(DocumentConst::DEFAULT_BUFFER_SIZE);
    filename = path;

    using std::ifstream;
    ifstream f(path);
    
    if (!f)
        throw std::invalid_argument("file did not open.");
    
    char c;
    while (f.get(c)) 
    {
        chstr.push_back(c);
        formatChars.push_back(Format::charColor::White);
    }
}